pints/_transformation.py

```diff
@@ -0,0 +1,1120 @@
+        result with respect to the parameters.
+
+        The returned data is a tuple ``(S, S')`` where ``S`` is a scalar value
+        and ``S'`` is a sequence of length ``n_parameters``.
+
+        Note that the derivative returned is of the log of the determinant of
+        the Jacobian, so ``S' = d/dq log(|det(J(q))|)``, evaluated at input.
+
+        *This is an optional method.* It is needed when transformation is
+        performed on :class:`LogPDF` and that requires ``evaluateS1()``.
+        """
+        # Directly calculate the derivative using jacobian_S1()
+        #
+        # d/dq log(|det(J(q))|) = 1/|det(J(q))| * sign(det(J(q)))
+        #                         * d/dq det(J(q))
+        #
+        # The second term is given by Eq. 46 in the Matrix Cookbook (2012)
+        # http://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf
+        #
+        # d/dq det(J(q)) = det(J) Tr(J^{-1} d/dq J)
+        #
+        # Therefore
+        #
+        # d/dq log(|det(J(q))|) = Tr(J^{-1} d/dq J)
+        #
+        q = pints.vector(q)
+        jac, jac_S1 = self.jacobian_S1(q)
+        out_S1 = np.zeros(q.shape)
+        for i, jac_S1_i in enumerate(jac_S1):
+            out_S1[i] = np.trace(np.matmul(np.linalg.pinv(jac), jac_S1_i))
+        return self.log_jacobian_det(q), out_S1
+
+    def multiple_to_model(self, qs):
+        """
+        Transforms a series of parameter vectors ``qs`` from the search space
+        to the model space. ``qs`` must be provided in the shape
+        (``n_vectors``, ``n_parameters``).
+        """
+        qs = np.asarray(qs)
+        ps = np.zeros(qs.shape)
+        for i, q in enumerate(qs):
+            ps[i, :] = self.to_model(q)
+        return ps
+
+    def n_parameters(self):
+        """
+        Returns the dimension of the parameter space this transformation is
+        defined over.
+        """
+        raise NotImplementedError
+
+    def to_model(self, q):
+        """
+        Transforms a parameter vector ``q`` from the search space to the model
+        space.
+        """
+        raise NotImplementedError
+
+    def to_search(self, p):
+        """
+        Transforms a parameter vector ``p`` from the model space to the search
+        space.
+        """
+        raise NotImplementedError
+
+
+class ElementWiseTransformation(Transformation):
+    r"""
+    Abstract base class for objects that provide *element-wise* transformations
+    between two parameter spaces: the model parameter space and a search space.
+    That is, for a parameter vector :math:`\boldsymbol{p}` in the model space
+    and a parameter vector :math:`\boldsymbol{q}` in the search space, the
+    transformation :math:`\boldsymbol{f}` relating the two vectors takes the
+    form
+
+    .. math::
+        q_i = f_i(p_i),
+
+    where :math:`x_i` denotes the :math:`i^{\text{th}}` element of the vector
+    :math:`\boldsymbol{x}`.
+    """
```

Comment:
Guess this can implement n_parameters() then?

Reply:
And can any of the other implemented methods be implemented faster/easier in the 1d case?

Reply:
What do you mean by "can implement n_parameters()"? n_parameters is specified in each class that extends this. This is not necessary to be 1d.

Reply:
What do you mean by "element-wise" if this is not for a single parameter (element) ?

Reply:
If this can work on any number of parameters, then I don't understand the need to define it. Surely you can compose transformations of any other transformations, including already composed ones (nesting) ?

Reply:
Hmm, I didn't explain that very well. I will try make it a bit more explicit (longer) here, and see what you think first.

Element-wise means the transformation applies to each _element_ of the parameter vector _independently_ (hence the name element-wise). So if we have a transformation `g`, then it does `q_i = g(p_i)` for every element `i` in the parameter vector. It is different from the general transformation in the sense that the transformation could do something like q1 = p1 + p2 and q2 = p1 * p2, which is a valid transformation but not element-wise. That is the whole idea of separating element-wise and the general one. This is done similarly in PyMC3!

Why not just for one element? Because often we would say just log transform the full parameter vector. Then it can be done simply with `LogTransformation(n_parameters=123)` rather than composing one transformation at a time...!

Reply:
Right. It applies _mathematically_ to each element independently, but in practice it does n of them in one go :D 

Where do we use the independent/mixed difference in our code?

Reply:
Is it between the ComposedTransformation and the ComposedElementWiseTransformations ? Probably best to cross-reference those then in their docstrings! So the composed one should say, if you're using all element-wise ones then using the other class if faster.

(Looking at them, they should also use the 1-line+space+paragraph docstring format)

Alternatively, you could have element_wise() as a method (or a private property), and a single ComposedTransformation that decides at construction time whether it is element wise or not, and then uses the appropriate calculations?

Reply:
Ha, if "mathematically" helps, yeah. :sweat_smile: 

No mixed one so far, all element-wise. But in case if anyone or later we want to have a mixed one.

Reply:
OK, after some more discussion with @MichaelClerx, I will move "element-wise" from subclass to a class property. So each Transformation class need one of
```
    def elementwise(self):
        """ Returns True if it is an element-wise transformation. """
        return NotImplementedError
```
And for the ComposedTransformatioin, which needs both, I will do something like
```
    def jacobian(self, q):                                                      
        """ See :meth:`Transformation.jacobian()`. """                          
        if self._elementwise:                                                   
            return self._elementwise_jacobian(q)                                
        else:                                                                   
            return self._general_jacobian(q)
```